### PR TITLE
Remove NumPy 2 pin, fix direct colormap with keys outside data dtype

### PR DESCRIPTION
**Pull request by jni**
Now that vispy 0.14.3 is released with numpy 2 compatibility, we should be able to support numpy 2 also.

Removing the pin and testing in numpy 2 exposed a bug in the DirectLabelColormap implementation: a direct colormap with keys outside of the range of the data dtype would silently overflow and overwrite earlier labels. This PR fixes it by ignoring values outside of the dtype range.

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
Good to keep an eye here:
https://github.com/numpy/numpy/issues/26191

My quick glance id's `imageio` as a potential problem

**Comment by Czaki** [comment]
> My quick glance id's `imageio` as a potential problem

So it means that we should ship a bundle with numpy<2 for 0.5.0 release?

**Comment by jni** [comment]
Yeah I'm curious — did the tests pass with imageio and numpy 2 before? (ie do we depend on the failing imageio behaviour in our tests?) Anyway, we'll find out soon enough I guess. 😅 

PS: It looks like I still have merge rights if someone wants to make a PR to imageio fixing it for np2 😂 (but for now it's bedtime for me...)

**Review by Czaki — APPROVED** [review]

**Comment by Czaki** [comment]
I think that you could merge it once CI pass.

> Yeah I'm curious — did the tests pass with imageio and numpy 2 before? (ie do we depend on the failing imageio behaviour in our tests?) Anyway, we'll find out soon enough I guess. 😅

We test only our interface to `imageio`, we do not test all napari use cases for it.

**Comment by Czaki** [comment]
Ok. I have constraints to numpy==2.0.0, to test if it works.

**Comment by aganders3** [comment]
@Czaki thanks for chasing all these down...tedious 😓 

I think this `scikit-image` will need to be bumped as well:
https://github.com/napari/napari/blob/aacec38c106a408a768a7fab984b9a7b8dee8a2c/resources/constraints/constraints_py3.9.txt#L462

**Comment by Czaki** [comment]
@napari-bot update constraints

**Comment by Czaki** [comment]
@napari-bot update constraints

**Comment by Czaki** [comment]
Test are failing because of tensorstore (I do not understand why 3.9 passes). 


What do we want to do? Remove tensorstore from default test dependencies?

**Comment by aganders3** [comment]
I think 3.9 passes because it's still testing with numpy 1.26.4 (due at least in part to the scikit-image constraint).

tensorstore fixes have been committed to their main branch, so hopefully a new release is coming shortly.
https://github.com/google/tensorstore/issues/165
https://github.com/google/tensorstore/commits/master/

**Comment by psobolewskiPhD** [comment]
@napari-bot update constraints

**Comment by jni** [comment]
Updated to ignore xarray deprecation warnings about copy keyword argument in `__array__`. See:

[failure in our GHA run](https://github.com/napari/napari/actions/runs/9557110080/job/26343640898?pr=6998#step:12:3838):

```
  scale = <xarray.DataArray (dim_0: 2)> Size: 16B
  array([1, 1])
  Dimensions without coordinates: dim_0
  
      def scale_to_vector(scale, *, ndim):
          """Convert a scale input into an n-dimensional transform component.
      
          Parameters
          ----------
          scale : 1-D array
              A 1-D array of factors to scale each axis by. Scale is padded with 1
              in leading dimensions, so that, for example, a scale of [4, 18, 34] in
              3D can be used as a scale of [1, 4, 18, 34] in 4D without modification.
              An empty vector implies no scaling.
          ndim : int
              The desired dimensionality of the output transform component.
      
          Returns
          -------
          np.ndarray
              The scale component as a 1D numpy array of length ndim.
          """
          scale_arr = np.ones(ndim)
          if scale is not None:
  >           scale_arr[-len(scale) :] = scale
  E           DeprecationWarning: __array__ implementation doesn't accept a copy keyword, so passing copy=False failed. __array__ must implement 'dtype' and 'copy' keyword arguments.
  
  napari/utils/transforms/transform_utils.py:138: DeprecationWarning
```

Corresponding xarray PR: pydata/xarray#8939

Corresponding lines in xarray pytest config: https://github.com/pydata/xarray/blame/b1f3fea467f9387ed35c221205a70524f4caa18b/pyproject.toml#L333-L334

**Comment by jni** [comment]
Cool, just the benchmarks to fix! 🥳 I am running them locally now. They are sloooooow far out. 😂

**Comment by Czaki** [comment]
Run all benchmarks in the fastest possible way 

```
PR=1 asv run --show-stderr --attribute timeout=30  --durations 5 --quick --python=same
```

run selected group of benchmarks 

```
PR=1 asv run --show-stderr --attribute timeout=30  --durations 5 --quick --python=same  -b benchmark_python_layer 
```

Where `-b` argument provides regexp pattern for selecting benchmark based on name.

**Comment by jni** [comment]
Ok the issue is in this line:

https://github.com/napari/napari/blob/05b8d88e8f059675abb1c44260060bb124fac82e/napari/utils/colormaps/colormap.py#L644

when the input dictionary contains values outside of the dtype range, we just ignore it. Actually that was probably a bug before that NumPy 2 has found: we would have been overflowing and overwriting the lower-valued colors with new ones. Which is nice, we can label this PR a bugfix also. 😂 Fixing!

**Comment by jni** [comment]
Oh no, should have read the [docstring](https://github.com/napari/napari/blob/05b8d88e8f059675abb1c44260060bb124fac82e/napari/utils/colormaps/colormap.py#L609-L618) before posting, that's not what this function is doing. It's mapping uint8 values to smaller uint8 values that correspond to the unique colors in the colormap. I think that might be wrong in a different way — I thought we decided not to remap the texture values if our data is only uint8...? 🤔

**Comment by jni** [comment]
At any rate, it is correct to ignore dictionary keys above the data type, since they could never appear in the data. So I'll do that.

**Comment by jni** [comment]
Ok I think this should be ready! 🥳 🚀

File: napari/_tests/test_dtypes.py [hunk 1]
```diff
@@ -12,7 +12,16 @@
     np.dtype(np.int32),
     np.dtype(np.uint32),
     np.dtype(np.int64),
-    np.dtype(np.uint64),
+    pytest.param(
```
Comment (aganders3) [inline napari/_tests/test_dtypes.py:15]: Here's a simple suggestion to fix the test IDs that were bothering you - change the decorator on the test below:
```python
@pytest.mark.parametrize('dtype', dtypes, ids=str)
```

**Pull request merged**